### PR TITLE
Support WSTOPPED and WCONTINUED

### PR DESCRIPTION
**Pull request by StevenJiang1110**
This PR adds the `WSTOPPED` and `WCONTINUED` flags to wait4 and waitpid, as required by #2112.

These two flags are used to monitor changes in child _thread_ states (from running to stopped or from stopped to running). To support this behavior, I have also moved the stop and continue thread status from `Thread` to `PosixThread`.

TODO:
1. The current wait implementation is racy; wake events may be missed.
2. Implement the WNOWAIT test.
3. Replace the busy loop with Pauser to wait until the thread is no longer stopped.

### CODE REVIEW

File: kernel/src/thread/status.rs [hunk 1]
```diff
@@ -27,10 +27,6 @@ impl ThreadStatus {
     pub fn is_exited(&self) -> bool {
         *self == ThreadStatus::Exited
     }
-
-    pub fn is_stopped(&self) -> bool {
-        *self == ThreadStatus::Stopped
-    }
```
Comment (lrh2000) [inline kernel/src/thread/status.rs:33]: So should we remove `ThreadStatus`? It seems that `ThreadStatus::Init` may not be very useful and `ThreadStatus::Stopped` is no longer useful after this PR.

File: kernel/src/thread/task.rs [hunk 2]
```diff
@@ -88,15 +90,71 @@ pub fn create_new_user_task(
...
+                    status = Status::Exit;
+                    break;
+                }
+
+                if !current_posix_thread.status().lock().is_stopped() {
+                    break;
+                }
+
+                // Slow path: The thread is stopped.
+                let wait_status = stop_wait_queue.pause_until(|| {
```
Comment (lrh2000) [inline kernel/src/thread/task.rs]: I don't think we need a wait queue. Maybe we can just use `Waiter::pause_until` here?

File: kernel/src/thread/task.rs [hunk 3]
```diff
@@ -88,15 +90,71 @@ pub fn create_new_user_task(
...
+                /// wake up the thread.
+                RecheckSignal,
             }
-            if current_thread.is_exited() {
-                debug!("exit due to signal");
+
+            let mut status = Status::Run;
+
+            loop {
+                // Fast path: Check if the thread has exited or is not stopped
```
Comment (lrh2000) [inline kernel/src/thread/task.rs]: `WaitQueue::pause_until` already has a fast path so there is no need to have another.

(The comment above suggests to use `Waiter::pause_until`, which does not have a fast path, though.)

File: kernel/src/thread/task.rs [hunk 4]
```diff
@@ -88,15 +90,71 @@ pub fn create_new_user_task(
...
+                let wait_status = stop_wait_queue.pause_until(|| {
+                    // FIXME: Should signals other than SIGCONT and SIGKILL be handled while the thread is stopped?
+                    handle_pending_signal(user_ctx, &ctx, syscall_number);
+
+                    if current_thread.is_exited() {
+                        return Some(Status::Exit);
+                    }
+
+                    // Check for pending signals before acquiring the thread status lock,
+                    // as `has_pending` may internally acquire a Mutex lock.
```
Comment (lrh2000) [inline kernel/src/thread/task.rs]: `pause_until` already checks pending signals. I'm not sure why we need this check here?

File: kernel/src/thread/task.rs [hunk 5]
```diff
@@ -88,15 +90,71 @@ pub fn create_new_user_task(
...
+                /// The thread is stopped, but there are pending signals.
+                /// We need to handle the pending signals to determine if any can
+                /// wake up the thread.
+                RecheckSignal,
             }
-            if current_thread.is_exited() {
-                debug!("exit due to signal");
+
+            let mut status = Status::Run;
+
```
Comment (lrh2000) [inline kernel/src/thread/task.rs]: Summarizing several comments above, I think the following simplified version works:
```rust
            loop {
                handle_pending_signal(user_ctx, &ctx, syscall_number);

                if current_thread.is_exited() {
                    break;
                }

                if !current_posix_thread.status().lock().is_stopped() {
                    break;
                }

                let _ = stop_waiter.pause_until(|| None);
            }
```

File: kernel/src/process/wait.rs [hunk 6]
```diff
@@ -96,16 +101,84 @@ pub fn wait_child_exit(
...
+    if wait_options.contains(WaitOptions::WNOWAIT) {
+        Some(wait_status)
+    } else {
+        let zombie_pid = zombie_child.pid();
+        reap_zombie_child(zombie_pid, children_lock);
+        Some(wait_status)
+    }
+}
+
+fn wait_thread(
```
Comment (lrh2000) [inline kernel/src/process/wait.rs]: I don't believe that any of `SIGSTOP`, `SIGCONT`, `WSTOPPED`, and `WCONTINUED` are related to threads. They affect the entire process.

Similarly, you cannot SIGKILL a single thread. It will kill the entire process.

File: kernel/src/process/wait.rs [hunk 7]
```diff
@@ -96,16 +101,84 @@ pub fn wait_child_exit(
...
+    unwaited_children: &[Arc<Process>],
+    wait_options: WaitOptions,
+    _children_lock: &mut BTreeMap<Pid, Arc<Process>>,
+) -> Option<WaitStatus> {
+    // Lock order: Children of process -> Tasks of process ->
```
Comment (lrh2000) [inline kernel/src/process/wait.rs]: Incomplete comments.

**Review by lrh2000 — COMMENTED** [review]
> These two flags are used to monitor changes in child thread states (from running to stopped or from stopped to running). 

Sources? AFAIK, you cannot SIGSTOP a single thread without stopping the entire process, just like you cannot SIGKILL a single thread without killing the entire process.

**Comment by StevenJiang1110** [comment]
> Sources? AFAIK, you cannot SIGSTOP a single thread without stopping the entire process, just like you cannot SIGKILL a single thread without killing the entire process.

Thanks for point out this! I misunderstood the implementation of [wait_task_stopped](https://elixir.bootlin.com/linux/v6.0.9/source/kernel/exit.c#L1142) in Linux. The `task` referenced here should refer to a process rather than a thread. I have conducted local experiments to confirm this behavior. I will correct implementation accordingly.

File: kernel/src/thread/task.rs (hunk 5 above)
Comment (StevenJiang1110) [reply kernel/src/thread/task.rs]: Since the stop status is maintained in the process now, and the SIGCONT will only be processed by one thread, other stopped threads won't be woken up or interrupted by SIGCONT. So I make the following 2 changes:

When resuming a process, it should wake up all threads of the process.

```rust
impl Process {
     /// Resumes the stopped process.
    pub fn resume(&self) {
        if self.status.stop_status().lock().resume() {
            self.wake_up_parent();

            // Note that the resume function is called by the thread which deals with SIGCONT,
            // since SIGCONT is handled by any thread in this process, we need to wake
            // up other stopped threads in the same process.
            for task in self.tasks.lock().as_slice() {
                let posix_thread = task.as_posix_thread().unwrap();
                posix_thread.wake_signal_waker();
            }
        }
    }
}
```

The stopped thread will wait until the process stop status changed:
```rust
loop {
    // Handle signal, check exited and stopped.

    // Wait until not stopped
    match stop_waiter.pause_until(|| (!current_process.is_stopped()).then_some(())) {
           // The process is already woken up by SIGCONT.
           Ok(()) => break,
           // The thread is interrupted by some signals.
           Err(_) => continue,
      }
}
```

Indeed, the current handling of SIGCONT may be incorrect (this issue existed prior to this PR). According to [a Zhihu documentation](https://zhuanlan.zhihu.com/p/587496380), SIGCONT will not be enqueued to the signal queues; instead, it is intended to wake up stopped threads. I have not yet checked the Linux code regarding this behavior.

```plain
SIGCONT信号的处理只有发送，没有deliver过程。
信号的发送其实只是唤醒那些处于TASK_STOPPED状态的线程， 
并没有将信号写入task的信号pending队列。
既然没有将信号写入task的pending队列，
也就没有线程对于SIGSCONT信号的deliver过程。
```

Comment (lrh2000) [reply kernel/src/thread/task.rs]: > Indeed, the current handling of SIGCONT may be incorrect (this issue existed prior to this PR). According to [a Zhihu documentation](https://zhuanlan.zhihu.com/p/587496380), SIGCONT will not be enqueued to the signal queues; instead, it is intended to wake up stopped threads. I have not yet checked the Linux code regarding this behavior.

That's interesting. Will it cause any behavioral differences? I think what you said is just an implementation detail?

Comment (StevenJiang1110) [reply kernel/src/thread/task.rs]: > Will it cause any behavioral differences?

I guess the difference is that SIGCONT should not interrupt blocking syscalls (like poll) on Linux, while it currently does interrupt blocking syscalls on Asterias. Alghouth I haven't checked it now.

File: kernel/src/process/posix_thread/mod.rs [hunk 8]
```diff
@@ -199,6 +199,13 @@ impl PosixThread {
         *self.signalled_waker.lock() = None;
     }
 
+    /// Wakes up the signal waker.
+    pub fn wake_signal_waker(&self) {
```
Comment (lrh2000) [inline kernel/src/process/posix_thread/mod.rs]: `wake_signalled_waker`? It's the name used in the `{set,clear}_*` methods.

File: test/apps/process/wait4.c [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+
+FN_TEST(continue_child)
+{
+	TEST_SUCC(kill(pid, SIGCONT));
+	status = 0;
+	TEST_RES(wait4(pid, &status, WCONTINUED, NULL),
+		 _ret == pid && WIFCONTINUED(status));
+	status = 0;
+	TEST_RES(wait4(pid, &status, WCONTINUED | WNOHANG, NULL),
+		 _ret == 0 && !WIFCONTINUED(status));
```
Comment (lrh2000) [inline test/apps/process/wait4.c]: It seems that you sometimes write:
```c
		 _ret == 0 && !WIFCONTINUED(status));
		 _ret == 0 && !WIFSTOPPED(status));
```
sometimes write:
```c
		 _ret == 0 && status == 0);
```

I think we can unify them to one style instead. Maybe just checking against 0.

File: kernel/src/thread/task.rs [hunk 10]
```diff
@@ -88,15 +90,29 @@ pub fn create_new_user_task(
...
             }
-            handle_pending_signal(user_ctx, &ctx, syscall_number);
-            // If current is suspended, wait for a signal to wake up self
-            while current_thread.is_stopped() {
-                Thread::yield_now();
-                debug!("{} is suspended.", current_posix_thread.tid());
-                handle_pending_signal(user_ctx, &ctx, None);
+
+            loop {
+                handle_pending_signal(user_ctx, &ctx, syscall_number);
```
Comment (lrh2000) [inline kernel/src/thread/task.rs]: Can you confirm that handling multiple nested signals, e.g., SIGSTOP, SIGINT, SIGTRAP, and SIGCONT, is fine here? At least, our implementation should be compatible with it. (I'm not sure about Linux's behavior, but we can study it in the future.)

File: kernel/src/thread/task.rs [hunk 11]
```diff
@@ -88,15 +90,29 @@ pub fn create_new_user_task(
...
+                    break;
+                }
+
+                if !current_process.is_stopped() {
+                    break;
+                }
+
+                match stop_waiter.pause_until(|| (!current_process.is_stopped()).then_some(())) {
+                    // The process is already woken up by SIGCONT.
+                    Ok(()) => break,
```
Comment (lrh2000) [inline kernel/src/thread/task.rs]: Even if it's `OK(())`, there can be signals as well?

File: kernel/src/thread/task.rs [hunk 12]
```diff
@@ -88,15 +90,29 @@ pub fn create_new_user_task(
...
-            handle_pending_signal(user_ctx, &ctx, syscall_number);
-            // If current is suspended, wait for a signal to wake up self
-            while current_thread.is_stopped() {
-                Thread::yield_now();
-                debug!("{} is suspended.", current_posix_thread.tid());
-                handle_pending_signal(user_ctx, &ctx, None);
+
+            loop {
+                handle_pending_signal(user_ctx, &ctx, syscall_number);
+                syscall_number = None;
```
Comment (lrh2000) [inline kernel/src/thread/task.rs]: Convert it to a `while` loop may be clearer:

```rust
    while !current_thread.is_exited() {
        // Execute the user code
        let return_reason = user_mode.execute(has_kernel_event_fn);

        // Handle user events
        let user_ctx = user_mode.context_mut();
        // ..

        // Exit if the thread terminates
        if current_thread.is_exited() {
            break;
        }

        // Handle signals
        handle_pending_signal(user_ctx, &ctx, syscall_number);

        // Handle signals while the thread is stopped
        while !current_thread.is_exited() && current_process.is_stopped() {
             let _ = stop_waiter.pause_until(|| (!current_process.is_stopped()).then_some(()));
             handle_pending_signal(user_ctx, &ctx, None);
        }
    }
```

File: kernel/src/process/status.rs [hunk 13]
```diff
@@ -71,3 +77,85 @@ impl ProcessStatus {
         self.exit_code.store(exit_code, Ordering::Relaxed);
     }
 }
+
+impl ProcessStatus {
+    pub(super) fn stop_status(&self) -> &SpinLock<StopStatus> {
```
Comment (lrh2000) [inline kernel/src/process/status.rs]: This PR adds a `SpinLock` operation to the `getpid` path. I suppose it has some impact on performance. Perhaps we could conduct some experiments or consider using atomic operations here (which seems a bit tricky with `has_status_changed`, though).

File: kernel/src/process/wait.rs [hunk 14]
```diff
@@ -96,16 +116,92 @@ pub fn wait_child_exit(
...
+fn reap_zombie_child(pid: Pid, children_lock: &mut BTreeMap<Pid, Arc<Process>>) -> ExitCode {
+    let child_process = children_lock.remove(&pid).unwrap();
     assert!(child_process.status().is_zombie());
 
     for task in child_process.tasks().lock().as_slice() {
         thread_table::remove_thread(task.as_posix_thread().unwrap().tid());
     }
 
-    // Lock order: session table -> group table -> process table -> group of process
+    // Lock order: children of process -> session table -> group table -> process table -> group of process
```
Comment (lrh2000) [inline kernel/src/process/wait.rs]: This line is too long. Maybe break it.

File: kernel/src/syscall/waitid.rs [hunk 15]
```diff
@@ -19,12 +19,12 @@ pub fn sys_waitid(
...
-    let waited_process =
-        wait_child_exit(process_filter, wait_options, ctx).map_err(|err| match err.error() {
+    let wait_status =
+        do_wait(process_filter, wait_options, ctx).map_err(|err| match err.error() {
             Errno::EINTR => Error::new(Errno::ERESTARTSYS),
             _ => err,
         })?;
 
-    let pid = waited_process.map_or(0, |process| process.pid());
+    let pid: u32 = wait_status.map_or(0, |wait_status| wait_status.pid());
```
Comment (lrh2000) [inline kernel/src/syscall/waitid.rs]: Why marking it as `u32`?

File: kernel/src/thread/task.rs (hunk 5 above)
Comment (lrh2000) [reply kernel/src/thread/task.rs]: OK. Thanks for your explanation. This sounds like a corner case. We can look into it in the future.

File: kernel/src/thread/task.rs (hunk 10 above)
Comment (StevenJiang1110) [reply kernel/src/thread/task.rs]: The nested signals can be handled, however, the current behavior is inconsistent with Linux.

Currently, we handle all possible signals when the process is stopped.
However, when the process is stopped, at least signals with user-provided handlers 
should not be handled; these signals should only be handled after the process is continued.
Certain signals, such as SIGKILL, should be handled even if the process is stopped.
We may need to further investigate Linux behavior regarding which signals should be handled when the thread is stopped.

I add a FIXME about this problem and a regression test about the behavior (some checks are commented out since it cannot pass on Asterinas).

File: kernel/src/process/status.rs (hunk 13 above)
Comment (StevenJiang1110) [reply kernel/src/process/status.rs]: I conducted the getpid test on my local SEV machine, and it shows that this PR introduces an average overhead of 2.4% in this benchmark. I guess this overhead may be acceptable?

(The results on my machine differ from those on the CI machine; the benchmark result of the main branch is only about half that of Linux on my machine.)

lmbench-getpid/us | With this PR | The main branch | Linux
-- | -- | -- | --
Tries 1 | 0.0960 | 0.094 | 0.0555
Tries 2 | 0.1006 | 0.0982 | 0.0555
Tries 3 | 0.1007 | 0.0983 | 0.0555
Average | 0.0991 | 0.0968 | 0.0555

> consider using atomic operations here (which seems a bit tricky with has_status_changed, though).

I also explored whether we could store `has_status_changed` and the stop/continue status in a single atomic variable. However, it doesn't seem easy to modify the variable in a single atomic instruction. Perhaps [fetch_update](https://doc.rust-lang.org/std/sync/atomic/struct.AtomicU32.html#method.fetch_update) could meet the requirements, but it may not execute in one atomic instruction.

Comment (lrh2000) [reply kernel/src/process/status.rs]: > I conducted the getpid test on my local SEV machine, and it shows that this PR introduces an average overhead of 2.4% in this benchmark. I guess this overhead may be acceptable?

Thanks for the experiments! It seems acceptable, but the system call path is very sensitive to performance because it affects the performance of all system calls. Therefore, it may be worth discussing this point further.

> However, it doesn't seem easy to modify the variable in a single atomic instruction. Perhaps [fetch_update](https://doc.rust-lang.org/std/sync/atomic/struct.AtomicU32.html#method.fetch_update) could meet the requirements, but it may not execute in one atomic instruction.

It uses compare-and-exchange. So semantically it should be correct. I'm not sure what's your concern about "not in a single atomic instruction"? Do you want to claim it's inefficient or incorrect?

---

Or maybe something like:
```rust
pub(super) struct StopStatus {
    is_stopped: AtomicBool,
    wait_status: SpinLock<Option<StopWaitStatus>>,
}

impl StopStatus {
    fn resume(&self, signum: SigNum) {
        // Hold the lock first to avoid race conditions
        let mut wait_status = self.wait_status.lock();

        if self.is_stopped.load(Ordering::Relaxed) {
            self.is_stopped.store(false, Ordering::Relaxed);
            *wait_status = Some(StopWaitStatus::Continue);
            true
        } else {
            false
        }
    }

    fn resume(&self, signum: SigNum) {
        // Hold the lock first to avoid race conditions
        let mut wait_status = self.wait_status.lock();

        if self.is_stopped.load(Ordering::Relaxed) {
            false
        } else {
            self.is_stopped.store(true, Ordering::Relaxed);
            *wait_status = Some(StopWaitStatus::Stopped(signum));
            true
        }
    }
}
```

So in the system call's fast path, we only need to check the `is_stopped` atomic boolean.

File: kernel/src/process/wait.rs [hunk 16]
```diff
@@ -96,16 +116,93 @@ pub fn wait_child_exit(
...
     assert!(child_process.status().is_zombie());
 
     for task in child_process.tasks().lock().as_slice() {
         thread_table::remove_thread(task.as_posix_thread().unwrap().tid());
     }
 
-    // Lock order: session table -> group table -> process table -> group of process
+    // Lock order: children of process -> session table
+    // -> group table -> process table -> group of process
     // -> group inner -> session inner
```
Comment (lrh2000) [inline kernel/src/process/wait.rs]: I think we can make them into two lines?
```suggestion
    // Lock order: children of process -> session table -> group table
    // -> process table -> group of process -> group inner -> session inner
```

File: kernel/src/thread/task.rs (hunk 10 above)
Comment (lrh2000) [reply kernel/src/thread/task.rs]: OK. Thanks for the explanation.

I'm surprised to learn that `handle_pending_signal` will process signals with user handlers immediately instead of waiting for the next user loop.

This does not sound right to me. However, this is a separate issue that we may want to address in the future.

File: kernel/src/process/status.rs (hunk 13 above)
Comment (StevenJiang1110) [reply kernel/src/process/status.rs]: The suggested code looks reasonable! Refactored as suggested.

> It uses compare-and-exchange. So semantically it should be correct. I'm not sure what's your concern about "not in a single atomic instruction"? Do you want to claim it's inefficient or incorrect?

Because I didn't understand the argument in the doc "This may call the function multiple times if the value has been changed from other threads in the meantime, as long as the function returns Some(_)", it makes me feel it will be called atomic instructions multiple times.

File: kernel/src/process/status.rs [hunk 17]
```diff
@@ -71,3 +77,97 @@ impl ProcessStatus {
...
+    }
+
+    /// Returns whether the process is stopped.
+    pub(super) fn is_stopped(&self) -> bool {
+        self.is_stopped.load(Ordering::Relaxed)
+    }
+
+    /// Waits on the stop status changes.
+    pub(super) fn wait(&self, options: WaitOptions) -> Option<StopWaitStatus> {
+        // Hold the lock first to avoid race conditions
```
Comment (lrh2000) [inline kernel/src/process/status.rs]: I don't think there are any race conditions? This method does not access/modify `is_stopped`.

File: kernel/src/process/status.rs [hunk 18]
```diff
@@ -71,3 +77,97 @@ impl ProcessStatus {
...
+            false
+        }
+    }
+
+    /// Returns whether the process is stopped.
+    pub(super) fn is_stopped(&self) -> bool {
+        self.is_stopped.load(Ordering::Relaxed)
+    }
+
+    /// Waits on the stop status changes.
```
Comment (lrh2000) [inline kernel/src/process/status.rs]: I don't think this method actually waits?

```suggestion
    /// Gets and clears the stop status changes for the `wait` syscall.
```

The same for the doc comments of `wait_stopped_or_continued`.

File: kernel/src/process/wait.rs [hunk 19]
```diff
@@ -31,26 +32,49 @@ bitflags! {
...
+            ctx.process.children_wait_queue().pause_until(|| {
+                // Acquire the children lock at first to prevent race conditions.
+                // We want to ensure that multiple waiting threads
+                // do not return the same waited process status.
+                let mut children_lock = ctx.process.children().lock();
+
+                let unwaited_children = children_lock
                     .values()
                     .filter(|child| match child_filter {
                         ProcessFilter::Any => true,
```
Comment (lrh2000) [inline kernel/src/process/wait.rs:79]: Maybe remove the following `cloned()` as well. It does not seem necessary since we hold the children lock in the entire closure.

File: test/apps/process/wait4.c [hunk 20]
```diff
@@ -0,0 +1,335 @@
...
+
+	// SIGKILL
+	TEST_SUCC(kill(child_pid, SIGKILL));
+	TEST_RES(wait4(child_pid, NULL, 0, NULL), _ret == child_pid);
+}
+END_TEST()
+
+FN_SETUP(kill_stopped)
+{
+	kill(pid, SIGKILL);
```
Comment (lrh2000) [inline test/apps/process/wait4.c]:
```suggestion
	CHECK(kill(pid, SIGKILL));
```

File: kernel/src/process/posix_thread/mod.rs [hunk 21]
```diff
@@ -199,6 +199,13 @@ impl PosixThread {
         *self.signalled_waker.lock() = None;
     }
 
+    /// Wakes up the signal waker.
```
Comment (lrh2000) [inline kernel/src/process/posix_thread/mod.rs]:
```suggestion
    /// Wakes up the signalled waker.
```

**Review by lrh2000 — COMMENTED** [review]
Just some minor nits, otherwise LGTM.

File: kernel/src/process/status.rs (hunk 13 above)
Comment (lrh2000) [reply kernel/src/process/status.rs]: > Because I didn't understand the argument in the doc "This may call the function multiple times if the value has been changed from other threads in the meantime, as long as the function returns Some(_)", it makes me feel it will be called atomic instructions multiple times.

It will retry if the atomic variable is updated during the closure. The closure cannot be executed atomically, but if the variable is updated in the middle, the compare-and-exchange after the closure will fail, so the it will try to execute the closure with the lastest value again.

If the atomic variable is not being contended, then it's just a corner case that should rarely occur.

File: test/apps/process/wait4.c [hunk 22]
```diff
@@ -0,0 +1,335 @@
...
+	TEST_RES(wait4(child_pid, &status, WSTOPPED | WNOHANG, NULL),
+		 _ret == 0);
+
+	TEST_SUCC(kill(child_pid, SIGKILL));
+	TEST_RES(wait4(child_pid, NULL, 0, NULL), _ret == child_pid);
+}
+END_TEST()
+
+int sigint_counter = 0;
+int sigtrap_counter = 0;
```
Comment (lrh2000) [inline test/apps/process/wait4.c]:
```
volatile int sigint_counter = 0;
volatile int sigtrap_counter = 0;
```

Variables that are accessed in the signal handlers should be volatile.

File: kernel/src/process/wait.rs [hunk 23]
```diff
@@ -96,17 +116,93 @@ pub fn wait_child_exit(
...
+    Stop(Arc<Process>, SigNum),
+    Continue(Arc<Process>),
+}
+
+impl WaitStatus {
+    pub fn pid(&self) -> u32 {
+        self.process().pid()
+    }
+
+    pub fn exit_status(&self) -> u32 {
```
Comment (lrh2000) [inline kernel/src/process/wait.rs]: Nowhere in the man pages (https://man7.org/linux/man-pages/man2/wait3.2.html) does it mention the term "exit status." I don't know where it comes from.

Perhaps the `exit_status_ptr` in `sys_wait4` should be renamed as `status_ptr`, and this method should be renamed as `fn status_code(self) -> u32`.

File: kernel/src/process/status.rs (hunk 13 above)
Comment (StevenJiang1110) [reply kernel/src/process/status.rs]: Thanks for your explanation!

File: kernel/src/process/status.rs [hunk 24]
```diff
@@ -71,3 +77,98 @@ impl ProcessStatus {
...
+        if options.contains(WaitOptions::WSTOPPED) {
+            let mut wait_status = self.wait_status.lock();
+
+            if let Some(StopWaitStatus::Stopped(_)) = wait_status.as_ref() {
+                return wait_status.take();
+            }
+        }
+
+        if options.contains(WaitOptions::WCONTINUED) {
+            let mut wait_status = self.wait_status.lock();
```
Comment (lrh2000) [inline kernel/src/process/status.rs]: Sorry, maybe I'm not clear in https://github.com/asterinas/asterinas/pull/2166#discussion_r2156666095. I just want to say that the comment is redundant since there are no race conditions to prevent. However, it seems like we still want to lock initially to avoid acquiring the lock twice if the user specifies `WSTOPPED | WCONTINUED`?

We have already checked one of them must present:
```rust
    if !wait_options.intersects(WaitOptions::WSTOPPED | WaitOptions::WCONTINUED) {
        return None;
    }
```
So here we don't need to worry the case where both of them are not present.

**Review by lrh2000 — APPROVED** [review]
Thanks for adding `WSTOPPED` and `WCONTINUED` support and for addressing the above comments. The changes look good to me.
